File: old/_14/go/_14_longest_common_prefix.go
package main

import "fmt"

// Write a function to find the longest common prefix string amongst an array of strings.

// If there is no common prefix, return an empty string "".

// Example 1:

// Input: ["flower","flow","flight"]
// Output: "fl"
// Example 2:

// Input: ["dog","racecar","car"]
// Output: ""
// Explanation: There is no common prefix among the input strings.
// Note:

// All given inputs are in lowercase letters a-z.
func longestCommonPrefix(strs []string) string {
    // fmt.Printf("strs: %v\n", strs)
    if len(strs) == 0 {
        return ""
    }
    i, res, isComplete := 0, "", false
    for (!isComplete) && i < len(strs[0]) {
        char := string(strs[0][i])
        // fmt.Println("char: ", char)
        for _, s := range strs[1:] {
            if i >= len(s) || string(s[i]) != char {
                isComplete = true
                return res
            }
        }
        res = res + char
        i = i+1
    }
    return res
}

func main() {
    testCases := [][]string{
        {"flower","flow","flight"},
        {"dog","racecar","car"},
        {"a"}
    }
    for _, test := range testCases {
        res := longestCommonPrefix(test)
        fmt.Println(res)
    }
}

File: old_session/session_1/_14/python/_14_longest_common_prefix.py
"""
14. Longest Common Prefix

Write a function to find the longest common prefix string amongst an array of strings.
"""
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        # Use strs[0] as basis, compare the first letter in rest elements,
        # if all same, compare the second and so on.
        # if find one differet, output the result
        if not strs: return ''
        i, res, isComplete = 0, '', False

        while not isComplete and i < len(strs[0]):      # make sure the index is correct
            char = strs[0][i]
            for s in strs[1:]:                          # compare the rest
                if i >= len(s) or s[i] != char:         # find one different (or missing)
                    isComplete = True
                    return res                          # output the result
            res += char
            i += 1

        return res